src/fsharp/CheckDeclarations.fs

```diff
@@ -2443,7 +2574,7 @@ let TcMutRecDefns_Phase2 (cenv: cenv) envInitial bindsm scopem mutRecNSInfo (env
                     // The rest should have been removed by splitting, they belong to "core" (they are "shape" of type, not implementation) 
                     | _ -> error(Error(FSComp.SR.tcDeclarationElementNotPermittedInAugmentation(), mem.Range))))
 
-
+      let extraBindings = Dictionary(HashIdentity.Reference)
```

Comment:
I'm a bit confused why the dictionary and two-phase approach is needed.  Can't we make the `PublishValueDefnMaybeInclCompilerGenerated ` calls in the first phase (the part with `Some tycon when declKind = DeclKind.ModuleOrMemberBinding`), and the `AugmentWithHashCompare.MakeBindingsForUnionAugmentation` in the second phase (the part fhtat comes after `MutRecBindingChecking.TcMutRecDefns_Phase2_Bindings`)?

Reply:
I mean, what you have is not wrong, and I can see that in order to do the second part you need to have the `vals` available from th first part anyway.  I guess ideally they would be stored in the MutRecShape. However what you have is ok, maybe put a comment in like this:

```
/// In the first phase of establishing the type definitions we create and publish the values for the `Is*` members. For convencience we also create the associated bindings. The bindings aren't actually needed until the second phase, after `MutRecBindingChecking.TcMutRecDefns_Phase2_Bindings`, so we store them away in a dictionary until they are needed.
```

Also, the dictionary should really be indexed by stamp (`tycon.Stamp`).  What is here is not wrong, it's just a little unusual in the compiler codebase to use object reference identity of `Tycon`/`Entity` as a key in a dictionary.  Not a big deal but might be best to adjust it.